### DISCUSSION TITLE
Colour

### DISCUSSION THREAD

**Discussion by KDMcMullan** [lhm0/rotating_display#2] [Ideas]
This is a nice piece of work really thoroughly documented. It's similar to something I've been paying with for a long time, although my design has a WeMos D1 and a LiPo cell running the parts which are necessarily mobile.

Have you considered (probably at a sacrifice to resolution) the possibility of a row of Red, Green and Blue LEDs with a view to creating full colour? My best guess is that the brightness control of each LED individually would be the limiting factor?

Well done, and thank you.

**Comment by lhm0** [comment]
Yes, I have thought about color quite a bit. 

There are several challenges involved. The most significant one is memory. I am using the Arduino nano, because it has the perfect timer for this application. However, the microcontroller has 2 kB of SRAM, only. This would definitely not be sufficient for color. Therefore, I am thinking about using a Raspberry pi pico, instead. RP pico has 256 kB SRAM - that would be perfect. Is there a version that has a smaller footprint than the RP pico?

The other challenge is the brightness control, indeed. I have an idea, on how to do it, but it would need a higher counting frequency. The nano is limited to 8MHz. I have not researched the RP pico yet and really know very little about it.

The "image generator" microcontroller, which is the ESP8266 right now, would also need to be upgraded. It has much larger memory, but still not sufficient for color images. 

The I2C link between the two processors would need to be replaced by SPI, because I2C is too slow.

I am looking for co-developers, particularly on the software side. My time available is so limited, unfortunately.

**Comment by KDMcMullan** [comment]
I don't know of a smaller RPi, but the Pico W would be a good choice (no Operating System). One thing I would say, however... There's an awful lot of your software which could be distributed to a different platform. I think your hardware selection is perfect, but it doesn't need to be a webserver. I feel like I would remove everything from the Arduino / ESP which is not required to be deterministic. I.e. remove everything but the time-critical display components and the bare minimum of comms, leaving it responsible only for rendering output.
The webserver, and what is ultimately displayed, then is the responsibility of a more computer-like device (eg a fullblown Raspberry Pi). 

(Though, it sounds, from the rest of your message, like the responsibility is already split up in that fashion.)

I have scheduled multiple PWM controllers on microcontrollers in assembler. Alongside the actual PWM percentage, there is a challenge of  ensuring that the rising edges coincide with the correct physical angle of rotation. I believe your program has a global brightness. Is that done by analogue voltage, or PWM? I haven't read that deeply.

I hear your challenge. I've done plenty of Arduino, but would never call myself a C programmer. Mostly Python recently. My interest would have been in developing the algorithm to translate the cartesian co-ordinate system to a rotary one!
 
Of course, I'll follow with interest and at least listen to bounce ideas off. Maybe even do some code!
